pam_kwallet.c

```diff
@@ -0,0 +1,823 @@
+    result = pam_set_data(pamh, kwalletPamDataKey, key, cleanup_free);
+
+    if (result != PAM_SUCCESS) {
+        free(key);
+        pam_syslog(pamh, LOG_ERR, "%s: Impossible to store the password: %s", logPrefix
+            , pam_strerror(pamh, result));
+        return PAM_IGNORE;
+    }
+
+    //TODO unlock kwallet that is already executed
+    return PAM_IGNORE;
+}
+
+static int drop_privileges(struct passwd *userInfo)
+{
+    /* When dropping privileges from root, the `setgroups` call will
+    * remove any extraneous groups. If we don't call this, then
+    * even though our uid has dropped, we may still have groups
+    * that enable us to do super-user things. This will fail if we
+    * aren't root, so don't bother checking the return value, this
+    * is just done as an optimistic privilege dropping function.
+    */
+    setgroups(0, NULL);
+
+    //Change to the user in case we are not it yet
+    if (setgid (userInfo->pw_gid) < 0 || setuid (userInfo->pw_uid) < 0 ||
+        setegid (userInfo->pw_gid) < 0 || seteuid (userInfo->pw_uid) < 0) {
+        return -1;
+    }
+
+    return 0;
+}
+
+static void execute_kwallet(pam_handle_t *pamh, struct passwd *userInfo, int toWalletPipe[2], char *fullSocket)
+{
+    //In the child pam_syslog does not work, using syslog directly
+
+    //keep stderr open so socket doesn't returns us that fd
+    int x = 3;
+    //Set FD_CLOEXEC on fd that are not of interest of kwallet
+    for (; x < 64; ++x) {
+        if (x != toWalletPipe[0]) {
+            fcntl(x, F_SETFD, FD_CLOEXEC);
+        }
+    }
+
+    //This is the side of the pipe PAM will send the hash to
+    close (toWalletPipe[1]);
+
+    //Change to the user in case we are not it yet
+    if (drop_privileges(userInfo) < 0) {
+        syslog(LOG_ERR, "%s: could not set gid/uid/euid/egit for kwalletd", logPrefix);
+        goto cleanup;
+    }
+
+    int envSocket;
+    if ((envSocket = socket(AF_UNIX, SOCK_STREAM, 0)) == -1) {
+        syslog(LOG_ERR, "%s: couldn't create socket", logPrefix);
+        goto cleanup;
+    }
+
+    struct sockaddr_un local = {};
+    local.sun_family = AF_UNIX;
+
+    if (strlen(fullSocket) > sizeof(local.sun_path)) {
+        syslog(LOG_ERR, "%s: socket path %s too long to open",
+                   logPrefix, fullSocket);
+        free(fullSocket);
+        goto cleanup;
+    }
+    strcpy(local.sun_path, fullSocket);
+    unlink(local.sun_path);//Just in case it exists from a previous login
+
+    syslog(LOG_DEBUG, "%s: final socket path: %s", logPrefix, local.sun_path);
+
+    if (bind(envSocket, (struct sockaddr *)&local, sizeof(local)) == -1) {
+        syslog(LOG_INFO, "%s-kwalletd: Couldn't bind to local file\n", logPrefix);
+        goto cleanup;
+    }
+
+    if (listen(envSocket, 5) == -1) {
+        syslog(LOG_INFO, "%s-kwalletd: Couldn't listen in socket: %d-%s\n", logPrefix, errno, strerror(errno));
+        goto cleanup;
+    }
+    //finally close stderr
+    close(2);
```

Comment:
Why close stderr at all? Every process is supposed to have a valid stderr stream. If this is to be discarded then `/dev/null` should be `dup2()'d` instead.